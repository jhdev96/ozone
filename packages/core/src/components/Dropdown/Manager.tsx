import React from 'react';
import ClickManager from '../ClickManager';
import OverflowManager from '../OverflowManager';


interface ManagerProps {
  /**
   * A ref object to pass to the ClickManager
   * component.
   */
  clickRef?: React.MutableRefObject<HTMLDivElement|undefined>;
  /**
   * A ref object to pass to the OverflowManager
   * component.
   */
  overflowRef?: React.MutableRefObject<HTMLDivElement|undefined>;
  /**
   * An optional css class that provides additional
   * styling.
   */
  className?: string;
  /**
   * A function that gets called when a user clicks
   * outside of the ClickManager component.
   */
  onClick?: () => void;
  /**
   * This is true when the dropdown is visible.
   */
  showDropdown: boolean;
  /**
   * A function which sets the state for the Dropdown
   * component.
   */
  setStateFn: (value: React.SetStateAction<boolean>) => void;
  /**
   * A React node around which this component wraps.
   */
  children?: React.ReactNode;
}

export default function Manager({
  clickRef,
  overflowRef,
  className,
  onClick,
  showDropdown,
  setStateFn,
  children
}: ManagerProps): JSX.Element 
{
  const localRef = overflowRef as React.Ref<HTMLDivElement>;

  return (
    <ClickManager 
      ref={clickRef} 
      className={className} 
      onClick={onClick}
    >
      <OverflowManager 
        ref={localRef} 
        isVisible={showDropdown}
        onOverflow={() => showDropdown && setStateFn(true)}
        onVisible={() => showDropdown && setStateFn(false)}
        updateParameters={[showDropdown]}
      >
        {children}
      </OverflowManager>
    </ClickManager>
  )
}