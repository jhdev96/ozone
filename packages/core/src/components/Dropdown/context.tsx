import React from 'react';


interface ContextType {
  /** The name of the selected item. */
  selected?: string;
  /** A function that handles selection. */
  handleSelection?: (name: string) => void;
};

export const DropdownContext = React.createContext<ContextType>({
  selected: '',
  handleSelection: () => {}
});