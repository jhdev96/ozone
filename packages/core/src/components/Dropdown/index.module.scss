@import '../../scss/root';
@import '../../scss/button.scss';


.basic,
.filter {
  min-width: 121px;
  width: fit-content;
  align-self: center;

  &__button {
    width: 100%;
    display: inline-flex;
    justify-content: space-between;
    padding: 5px 8px;
    box-sizing: border-box;
  }
}

.basic { margin-right: 10px; }

.basic__button {
  @extend .button--basic;

  margin: 0;
  border-radius: 0;
  padding: 8px 10px;
  cursor: pointer;

  &--light {
    color: var(--black-core);
    background-color: #fff;
    border: 1px solid var(--gray-3);
  }

  &--dark {
    color: var(--gray-1);
    border: 1px solid var(--black-6);
    background-color: var(--black-4) !important;
  }
}

.dropdown__menu--basic {
  margin-top: -1px !important;
  border-radius: 0 !important;

  :first-child {
    margin-top: 0 !important;
  }

  .menu__item {
    padding: 0.569rem !important;

    &:hover {background: var(--blue-0);}
  }

  &--light {
    color: #353640;
    border: 1px solid #dfe7ee;
    background-color: #fff;
  }
  &--dark {
    color: var(--gray-core);
    border: 1px solid #2b343e;
    background-color: #383f47 !important;
  }
}

.filter__button {
  position: relative;
  z-index: 999;
  min-width: 50px;
  border-radius: 100px;
  font-size: 0.9em;
  cursor: pointer;

  &--light {
    color: var(--black-2);
    background-color: var(--gray-core);
  }

  &--dark {
    color: var(--gray-3);
    background-color: #32383f;
  }

  &--visible {
    border-radius: 10px 10px 0 0 !important;
  }
}

.dropdown__menu--filter {
  z-index: 1;
  border-radius: 0 0 10px 10px !important;

  &:last-child { padding-bottom: 7px; }

  &--light {
    color: var(--black-2);
    background-color: var(--gray-core);

    .menu__item:hover {
      color: var(--black-core);
      background-color: transparent;
    }
  }

  &--dark {
    color: var(--gray-3);
    background-color: var(--black-core);

    .menu__item:hover {
      color: #fff;
      background-color: transparent;
    }
  }
}

.dropdown__menu {
  visibility: hidden;
  position: absolute !important;
  min-width: 121px;
  font-size: 0.8em;
  border-radius: 4px;
  box-sizing: border-box;
  transition: 0.2s height;

  .menu__item {
    padding: 0.25rem 0.65rem;
    text-align: left;
    cursor: pointer;
    transition: 0.2s opacity;
  }
}

.selected__title {
  color: #25aae1;
  font-weight: 600;
  text-shadow: none;
}

.visible {
  visibility: visible;
}

.aboveTarget {
  .dropdown__menu--basic {
    bottom: 32px;
  }

  .filter__button {
    border-radius: 0 0 10px 10px !important;
  }

  .dropdown__menu--filter {
    bottom: 25px;
    padding-top: 5px;
    border-radius: 10px 10px 0 0 !important;
  }
}

.selected { color: #25aae1 !important; }