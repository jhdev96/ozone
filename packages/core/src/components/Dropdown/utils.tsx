import React from 'react';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export const menuOptions = (
  selectedOption: string, 
  options: (string | React.ReactNode)[], 
  onClick: (e: React.MouseEvent<HTMLDivElement>) => void
): React.ReactNode => {
  /**
    * Return an array of div elements
    * based on the options given by 
    * the client. Each div should 
    * have an onClick prop specifying
    * the function to be invoked when
    * the div or its child element is
    * clicked.
    */
  const optionDivs = options.map((
    option: string | React.ReactNode, 
    idx: React.Key
  ) => (
    <div 
      onClick={onClick} 
      className={styles["menu__option"]} 
      key={idx}
    >
      <span
        className={classNames(
          option === selectedOption && styles.selected
        )}
      >
        {option}
      </span>
    </div>
  ));

  return optionDivs;
};