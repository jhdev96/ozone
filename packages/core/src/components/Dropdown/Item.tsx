import React, { 
  useContext, 
  useEffect, 
  useState,
} from 'react';
import classNames from '../../utils/classNames';
import { DropdownContext } from './context';
import styles from './index.module.scss';


interface ItemProps {
  /**
   * Sets the HTML `id` attribute for
   * the div element below.
   */
  id?: string;
  /**
   * A string value which will be passed
   * to a callback when the item is clicked. 
   * 
   * If a name is not provided, then the
   * item's textContent will be used instead.
   */
  name?: string;
  /**
   * The item's content.
   */
  children?: React.ReactNode;
}

export default function Item({
  id,
  name,
  children
}: ItemProps): JSX.Element 
{
  const {selected, handleSelection} = useContext(DropdownContext);
  const [isSelected, setIsSelected] = useState(false);
  const spanRef = React.createRef<HTMLSpanElement>();

  const selectThisItem = (e: React.MouseEvent<HTMLDivElement>) => {
    const text = e.currentTarget.textContent;
    let selection: string;

    if (name) {
      selection = name;
    } else if (text) { 
        selection = text; 
    } else { 
        selection = ''; 
    }

    handleSelection && handleSelection(selection);
  }

  useEffect(() => {
    if (spanRef.current) {
      const text = spanRef.current.textContent;

      if (name === selected || text === selected) {
        setIsSelected(true);
      } else {
          setIsSelected(false);
      }
    }
  }, [selected, spanRef.current]);

  return (
    <div 
      id={id}
      onClick={selectThisItem} 
      className={styles.menu__item} 
    >
      <span
        className={classNames(
          isSelected && styles.selected
        )}
        ref={spanRef}
      >
        {children}
      </span>
    </div>
  )
}

Item.displayName = 'DropdownItem';