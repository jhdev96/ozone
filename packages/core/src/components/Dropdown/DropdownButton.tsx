import React, { useContext } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import Colors from '@jhdev96/ozone-colors';
import { Arrow } from '@jhdev96/ozone-icons';
import { DropdownContext } from './context';
import styles from './index.module.scss';


interface DropdownButtonProps {
  /**
   * The type of dropdown being rendered.
   */
  variation?: "basic" | "filter";
  /**
   * Gives the button a light or dark theme.
   */
  theme?: ThemeType;
  /**
   * A function that is called when the button
   * is clicked.
   */
  onClick?: () => void;
  /**
   * Sets the direction of the arrow icon on
   * the button. This will be passed as a prop
   * to arrow icon.
   */
  arrowDirection?: "up" | "down";
  /**
   * This will be true when the dropdown is
   * visible.
   */
  showDropdown: boolean;
}

export default function DropdownButton({
  variation,
  theme,
  onClick,
  arrowDirection,
  showDropdown
}: DropdownButtonProps): JSX.Element 
{
  const {selected} = useContext(DropdownContext);

  return (
    <div
      onClick={onClick} 
      className={classNames(
        styles[`${variation}__button`], 
        styles[`${variation}__button--${theme}`],
        showDropdown && styles[`${variation}__button--visible`]
      )}
    >
      <span className={styles[`${variation}__title`]}>
        {selected}
      </span>
      <Arrow 
        direction={arrowDirection} 
        color={Colors.grayScale.BLACK2} 
      />
    </div>
  )
}