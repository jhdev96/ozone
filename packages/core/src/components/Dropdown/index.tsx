import React, { useState, useRef, LegacyRef } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import { DropdownContext } from './context';
import DropdownButton from './DropdownButton';
import Manager from './Manager';
import Item from './Item';
import styles from './index.module.scss';


interface DropdownProps {
  /** 
   * Sets the theme for the dropdown. 
   */
  theme?: ThemeType;
  /** 
   * Main classnames for dropdowns. 
   */
  variation?: "basic" | "filter";
  /** 
   * When true, this will display 
   * the selected option as the new title. 
   */
  saveSelection?: Boolean;
  /** 
   * Give the dropdown a default option when rendered. 
   */
  defaultValue?: string;
  /** 
   * The function to be invoked when a selection is made. 
   * A `selectedItem` parameter should be given to the
   * callback so that when an option is selected, its
   * value is received by the dropdown's parent element.
   */
  onSelection?: (selectedItem: string) => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /**
   * The dropdown menu items.
   */
  children?: React.ReactNode;
}

export default function Dropdown({
  theme           = "light", 
  variation       = "basic",
  saveSelection,
  defaultValue    = "Dropdown",
  onSelection     = () => {},
  dataTestId,
  children
}: DropdownProps): JSX.Element 
{
  const [selected, setSelected] = useState<string>(defaultValue);
  const [visible, setVisible] = useState(false);
  const [aboveTarget, setAboveTarget] = useState(false);
  const wrapperRef = useRef<HTMLDivElement>();
  const menuRef = useRef<HTMLDivElement>();

  const toggleDropdown = () => {
    setVisible(!visible);
    setAboveTarget(false);
  }

  const handleSelection = (name: string) => {
    /** 
     * display the selected option from the dropdown menu 
     * if the `saveSelection` prop is set to true.
     */
    if (saveSelection) setSelected(name); 
    
    name && onSelection(name);
    toggleDropdown();
  }
       
  return (
    <Manager
      clickRef={wrapperRef}
      overflowRef={menuRef}
      className={classNames(
        styles[variation],
        aboveTarget && styles.aboveTarget
      )}
      onClick={() => {visible && toggleDropdown()}}
      setStateFn={setAboveTarget}
      showDropdown={visible}
    >
      <DropdownContext.Provider value={{selected, handleSelection}}>
        <DropdownButton
          variation={variation}
          theme={theme}
          onClick={toggleDropdown}
          showDropdown={visible}
          arrowDirection={aboveTarget ? "up" : "down"}
        />
        <div 
          className={classNames(
            styles['dropdown__menu'],
            styles[`dropdown__menu--${variation}`],
            styles[`dropdown__menu--${variation}--${theme}`],
            visible && styles.visible
          )}
          ref={menuRef as LegacyRef<HTMLDivElement>}
          data-testid={dataTestId}
        >
          {children}
        </div>
      </DropdownContext.Provider>
    </Manager>
  )
}

Dropdown.displayName = 'Dropdown';
Dropdown.Item = Item;