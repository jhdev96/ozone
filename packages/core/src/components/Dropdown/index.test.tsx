import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import Dropdown from '.';


describe('<Dropdown />', () => {
  it('renders a `basic` dropdown by default', () => {
    render(
      <Dropdown dataTestId="test" />
    );

    expect(screen.getByTestId('test')).toMatchSnapshot();
  });

  it('renders a dropdown of type `filter`', () => {
    render(
      <Dropdown variation="filter" dataTestId="test" />
    );

    expect(screen.getByTestId('test')).toMatchSnapshot();
  });

  it('renders a dropdown with options', () => {
    render(
      <Dropdown dataTestId="test">
        <Dropdown.Item>A</Dropdown.Item>
        <Dropdown.Item>B</Dropdown.Item>
        <Dropdown.Item>C</Dropdown.Item>
      </Dropdown>
    );

    expect(screen.getByTestId('test').children)
      .toHaveLength(3);
  });

  it('renders a dropdown with a `defaultValue prop`', () => {
    const defaultValue = "Start";

    render(
      <Dropdown dataTestId="test" defaultValue={defaultValue} />
    );

    expect(document.querySelector('.basic__button'))
      .toHaveTextContent(defaultValue);
  })

  it('adds a visible class when the dropdown button is clicked', () => {
    render(
      <Dropdown dataTestId="test">
        <Dropdown.Item>A</Dropdown.Item>
        <Dropdown.Item>B</Dropdown.Item>
        <Dropdown.Item>C</Dropdown.Item>
      </Dropdown>
    );

    fireEvent.click(document.querySelector('.basic__button')!);
    expect(screen.getByTestId('test')).toHaveClass('visible');
  });

  it('changes the selection when an item is clicked', () => {
    render(
      <Dropdown saveSelection>
        <Dropdown.Item id="a">A</Dropdown.Item>
        <Dropdown.Item id="b">B</Dropdown.Item>
        <Dropdown.Item id="c">C</Dropdown.Item>
      </Dropdown>
    );

    fireEvent.click(document.querySelector('#a')!);
    expect(document.querySelector('.basic__button'))
      .toHaveTextContent('A');

    expect(document.querySelector('#a')!.firstChild)
      .toHaveClass('selected');

    fireEvent.click(document.querySelector('#b')!);
    expect(document.querySelector('.basic__button'))
      .toHaveTextContent('B');

    expect(document.querySelector('#b')!.firstChild)
      .toHaveClass('selected');

    fireEvent.click(document.querySelector('#c')!);
    expect(document.querySelector('.basic__button'))
      .toHaveTextContent('C');

    expect(document.querySelector('#c')!.firstChild)
      .toHaveClass('selected');
  });

  it('calls `onSelection` when an item is clicked', () => {
    const onSelect = jest.fn();

    render(
      <Dropdown onSelection={onSelect}>
        <Dropdown.Item id="a">A</Dropdown.Item>
        <Dropdown.Item>B</Dropdown.Item>
        <Dropdown.Item>C</Dropdown.Item>
      </Dropdown>
    );

    fireEvent.click(document.querySelector('#a')!);
    expect(onSelect).toBeCalledWith("A");
  });
});