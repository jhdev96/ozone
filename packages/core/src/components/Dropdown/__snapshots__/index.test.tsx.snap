// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders a \`basic\` dropdown by default 1`] = `
<div
  class="dropdown__menu dropdown__menu--basic dropdown__menu--basic--light"
  data-testid="test"
/>
`;

exports[`<Dropdown /> renders a dropdown of type \`filter\` 1`] = `
<div
  class="dropdown__menu dropdown__menu--filter dropdown__menu--filter--light"
  data-testid="test"
/>
`;
